### ISSUE TITLE
KOReader crashed with dump attempt to index a boolean value at reader.lua:23

### ISSUE THREAD

**Issue by TanyuT** [koreader/koreader#5589]
Hello,

I upgraded KOReader to 2019.11 from 2019.09 and it worked fine until I switched to nickel, once I returned to KOReader it crashed with dump:

./luajit: ./reader.lua:23: attempt to index a boolean value
stack traceback:
	./reader.lua:23: in main chunk
	[C]: at 0x000140c1

I cannot start it now, it always crashes.

I had similar issue with 2019.10 but I didn’t time to play with it and just reverted to 2019.09.

Kobo Clara, FW: 4.18.13737
KOReader 2019.11

Regards,
TT
[crash.log](https://github.com/koreader/koreader/files/3839002/crash.log)

**Comment by Frenzie** [comment]
@TanyuT How did you update? Could you please share the content of your `git-rev` file? It's crashing trying to write the version to the log. O_o

**Comment by TanyuT** [comment]
@Frenzie ,
Once I commented line 23, it dumped at line 29. I updated over the air. Everything went fine during the update. 
Attaching new crash log with line 23 commented out and git-rev.

[crash.log](https://github.com/koreader/koreader/files/3840867/crash.log)

[git-rev.zip](https://github.com/koreader/koreader/files/3840870/git-rev.zip)

**Comment by Frenzie** [comment]
Yeah, that confirms that it's not loading the modules.

Are they actually there, like in `.adds/koreader/frontend/version.lua`?

**Comment by TanyuT** [comment]
There a lot of lua files but all of them are with size 0. It seems this is the root cause? Shall I try to download the new version and install it manually?

**Comment by Frenzie** [comment]
That would certainly explain why it's not loading, but not why it happened. :-/ I use OTA update all the time myself.

Could you be (nearly) out of space by any chance? Say, less than 50 MB.

**Comment by Frenzie** [comment]
Oh, also, there should be a file in koreader/ota.

koreader.installed.tar

Actually it seems to be about 70 MB for me, so I guess you need at the very least about a 100 MB of free space.

Anyway, what does that file look like for you?

**Comment by NiLuJe** [comment]
You also appear to be using KSM, which is an unsupported setup on pretyt much any Mk. 7 ;).

**Comment by TanyuT** [comment]
@Frenzie , I have plenty of space, more than 6GB. The contents of the /ota is:
-rw-r--r-- 1 t t        0 2019-09-11 13:58 koreader.installed.tar
-rw-r--r-- 1 t t   397578 2019-03-20 17:35 koreader-kobo-latest-nightly.zsync
-rw-r--r-- 1 t t   425691 2019-11-12 02:58 koreader-kobo-latest-stable.zsync
-rw-r--r-- 1 t t 70000640 2019-05-08 16:34 koreader.updated.tar.zs-old
-rw-r--r-- 1 t t        0 2019-11-12 03:00 package.index

@NiLuJe , Yes, I use KSM, the version for Clara and it works(ed) pretty well so far.

**Comment by TanyuT** [comment]
@Frenzie , I have plenty of space, more than 6GB. The contents of the /ota is:
-rw-r--r-- 1 t t        0 2019-09-11 13:58 koreader.installed.tar
-rw-r--r-- 1 t t   397578 2019-03-20 17:35 koreader-kobo-latest-nightly.zsync
-rw-r--r-- 1 t t   425691 2019-11-12 02:58 koreader-kobo-latest-stable.zsync
-rw-r--r-- 1 t t 70000640 2019-05-08 16:34 koreader.updated.tar.zs-old
-rw-r--r-- 1 t t        0 2019-11-12 03:00 package.index

@NiLuJe , Yes, I use KSM, the version for Clara and it works(ed) pretty well so far.




I did it once more - manually installed 2019.09, tested it - it worked fine, updated, check the files in frontend (everything was OK), started KOReader successfully, switched to nickel and back to KOReader - this time it works. Also, noticed that the file koreader.installed.tar is 69MB. Am I right that the update file was damaged and I was the wrong impression it was updated while actually it wasn't and once it was restarted it failed?

**Comment by Frenzie** [comment]
The part before restarting is just the download & preparation of the update. If you delete the stuff in the `ota/` folder before starting the program with an update pending then nothing will happen.

The download creates a `koreader.updated.tar` file. If such a file is detected, it's extracted on top of the current installation, and subsequently moved to `koreader.installed.tar`. That latter file is used to compare differences with the latest update on the server, so it only has to download changes.

But if the download of the update fails, it removes all traces. It would be good to know how you ended up with an empty file. :-/

Anyway, we can probably include a simple sanity check. I'm not sure if `du` is the best option, but some facsimile of it is available in busybox.

```sh
# du koreader.installed.tar | cut -f 1
70700
```

So basically we could just say something like:

```sh
# Basic sanity check.
if [ $(du "${NEWUPDATE}"  | cut -f 1) -gt 60000 ]; then
  echo "onoez we is too small"
fi
```

**Comment by Frenzie** [comment]
I'm also not really sure how a 0-byte file could overwrite anything. I can only assume it was damaged somehow and didn't actually become 0-byte until the move operation?

Did you run a checkdisk on your partition? You can do this while it's mounted in Windows (if you use Windows).

**Comment by TanyuT** [comment]
I run checkdisk and everything was fine.

I will monitor it for some time. Is there an option to turn on verbose logging during update?

**Comment by Frenzie** [comment]
Any errors during OTA download should already be logged. Debug logging would only say where it's downloading the OTA file from.

**Comment by NiLuJe** [comment]
zsync does integrity checking during the download. Barring anything weird, it shouldn't be able to write a corrupt/incomplete koreader.updated.tar

(And the fact that both tarballs and the existing tree are here at all times make storage space issues extremely tricky to actually hit, too).

**Comment by Frenzie** [comment]
True, we're not talking about a Kindle here. (Given what you told me the other day in #5577 it doesn't sound all that tricky. ;-P)
